README.md
# climate-mood
Auto-updated hourly sentiment analysis for climate change on Twitter [Live Site](http://www.climatemood.io/)

## Main concepts
- `tweepy`: A python package used for querying Twitter.
The used query is `climate change -filter:retweets -filter:replies `
- `textblob`: A python package used for determining the subjectivity and polarity of tweets

## Dependencies
TODO

## Contributing guidelines
TODO


changes
diff --git a/app.py b/app.py
index 1896afd..f550310 100644
--- a/app.py
+++ b/app.py
@@ -129,8 +129,8 @@ app.layout = html.Div(children=[
                         'layout': go.Layout(
                                   title='Avg Characters per Tweet',
                                   hovermode=False,
-                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False),
-                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False),
+                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
+                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                   margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                                   width=300,
                                   height=300
@@ -152,8 +152,8 @@ app.layout = html.Div(children=[
                         'layout': go.Layout(
                                   title='Sum of Tweets per Mood',
                                   hovermode=False,
-                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False),
-                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False),
+                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
+                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                   margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                                   width=300,
                                   height=300,
@@ -177,8 +177,8 @@ app.layout = html.Div(children=[
                         'layout': go.Layout(
                                   title='Avg Subjectivity per Mood',
                                   hovermode=False,
-                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False),
-                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False),
+                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
+                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                   margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                                   width=300,
                                   height=300


climate_tweets.py
import os
import tweepy
from tweepy import OAuthHandler
import pandas as pd
import re
from textblob import TextBlob
import psycopg2
from sqlalchemy import create_engine

# Environment Variables to Twitter App
consumerKey = os.environ["CONSUMER_KEY"]
consumerSecret = os.environ["CONSUMER_SECRET"]
accessToken = os.environ["ACCESS_TOKEN"]
accessTokenSecret = os.environ["ACCESS_TOKEN_SECRET"]

# Environment Variables to Database
tweet_db_url = os.environ["DATABASE_URL"]

# Declaring relevant Variables
positive = 0
negative = 0
neutral = 0
polarityVal = []
subjectivityVal = []
tweetsArray = []
timestampArray = []
utfArray = []

def tweets_to_db():

        # Authenticating the Twitter Application and creating an object to use the Twitter API
        auth = tweepy.OAuthHandler(consumerKey, consumerSecret)
        auth.set_access_token(accessToken, accessTokenSecret)
        api = tweepy.API(auth, wait_on_rate_limit=True)

        #Get the tweets
        tweets = tweepy.Cursor(api.search, q='climate change -filter:retweets -filter:replies', lang='en').items(2000)
	
        # Creating a list of String Tweets
        for tweet in tweets:
            tweetsArray.append(tweet.text)
            timestampArray.append(tweet.created_at)
            utfArray.append(tweet.text.encode("utf-8"))
	
        # Cleaning a Tweet String
        def cleanTweet(tweet):
            # Remove Links, Special Characters etc from tweet
            return ' '.join(re.sub("([^0-9A-Za-z \t])|(\w+:\/\/\S+)"," ",tweet).split())

        # Perform Sentiment Analysis
        for tweet in tweetsArray:
            t = cleanTweet(tweet)
            analysis = TextBlob(t)
            polarityVal.append(analysis.sentiment.polarity)
            subjectivityVal.append(analysis.sentiment.subjectivity)

        # Create pandas dataframe from arrays
        tweet_frame = pd.DataFrame({"Tweets":tweetsArray, "TimeStamp":timestampArray, "UTF_Offset":utfArray, "Polarity":polarityVal, "Subjectivity":subjectivityVal}, columns=["Tweets", "TimeStamp", "UTF_Offset", "Polarity", "Subjectivity"])

        # Connect to Heroku and write to db
        engine = create_engine(tweet_db_url)
        tweet_frame.to_sql('tweets', con=engine, if_exists='replace')
	
        pass


tweets_to_db()


climate_tweets_rake.py
import os
import tweepy
from tweepy import OAuthHandler
import pandas as pd
import re
from textblob import TextBlob
import psycopg2
from sqlalchemy import create_engine

# Environment Variables to Twitter App
consumerKey = os.environ["CONSUMER_KEY"]
consumerSecret = os.environ["CONSUMER_SECRET"]
accessToken = os.environ["ACCESS_TOKEN"]
accessTokenSecret = os.environ["ACCESS_TOKEN_SECRET"]

# Environment Variables to Database
tweet_db_url = os.environ["DATABASE_URL"]

# Declaring relevant Variables
positive = 0
negative = 0
neutral = 0
polarityVal = []
subjectivityVal = []
tweetsArray = []
timestampArray = []
utfArray = []

def tweets_to_db():

        # Authenticating the Twitter Application and creating an object to use the Twitter API
        auth = tweepy.OAuthHandler(consumerKey, consumerSecret)
        auth.set_access_token(accessToken, accessTokenSecret)
        api = tweepy.API(auth, wait_on_rate_limit=True)

        #Get the tweets
        tweets = tweepy.Cursor(api.search, q='place:96683cc9126741d1 climate change -filter:retweets -filter:replies', lang='en').items(5000)
	
        # Creating a list of String Tweets
        for tweet in tweets:
            tweetsArray.append(tweet.text)
            timestampArray.append(tweet.created_at)
            utfArray.append(tweet.text.encode("utf-8"))
	
        # Cleaning a Tweet String
        def cleanTweet(tweet):
            # Remove Links, Special Characters etc from tweet
            return ' '.join(re.sub("([^0-9A-Za-z \t])|(\w+:\/\/\S+)"," ",tweet).split())

        # Perform Sentiment Analysis
        for tweet in tweetsArray:
            t = cleanTweet(tweet)
            analysis = TextBlob(t)
            polarityVal.append(analysis.sentiment.polarity)
            subjectivityVal.append(analysis.sentiment.subjectivity)

        # Create pandas dataframe from arrays
        tweet_frame = pd.DataFrame({"Tweets":tweetsArray, "TimeStamp":timestampArray, "UTF_Offset":utfArray, "Polarity":polarityVal, "Subjectivity":subjectivityVal}, columns=["Tweets", "TimeStamp", "UTF_Offset", "Polarity", "Subjectivity"])

        # Connect to Heroku and write to db
        engine = create_engine(tweet_db_url)
        tweet_frame.to_sql('tweets', con=engine, if_exists='replace')
	
        pass


tweets_to_db()


app.py
import os
import dash
from dash.dependencies import Input, State, Output
import dash_core_components as dcc
import dash_html_components as html
import pandas as pd
from pandas.tseries.offsets import *
import plotly.graph_objs as go
from numpy import arange,array,ones
from scipy import stats
import psycopg2
from sqlalchemy import create_engine
from datetime import datetime
from flask import Flask, send_from_directory

app = dash.Dash(__name__)
server = app.server
app.title = 'Climate Mood'

@server.route('/favicon.ico')
def favicon():
    return send_from_directory(os.path.join(server.root_path, 'assets'), 'favicon.ico')

tweet_db_url = os.environ['DATABASE_URL']

def get_df_from_db():

    engine = create_engine(tweet_db_url)

    df = pd.read_sql_table('tweets', engine)

    df['TimeStamp'] = pd.to_datetime(df['TimeStamp'], unit='s')
    df = df.set_index(df['TimeStamp'])
    df = df[~df.index.duplicated(keep='first')].sort_index()

    df = df.loc[df.iloc[df.index.get_loc((df.index[-1] - pd.DateOffset(hours=1)), method='nearest')]['TimeStamp']:df.index[-1]]

    return df

def sentiment_frames():

    sf = get_df_from_db()

    pf = sf[sf['Polarity'] > 0].groupby(pd.Grouper(freq='Min'))['Polarity'].count().fillna(0).to_frame()
    nf = sf[sf['Polarity'] < 0].groupby(pd.Grouper(freq='Min'))['Polarity'].count().fillna(0).to_frame()
    nef = sf[sf['Polarity'] == 0].groupby(pd.Grouper(freq='Min'))['Polarity'].count().fillna(0).to_frame()

    s_m = pd.merge(pf, nf, left_index=True, right_index=True)
    s_m = pd.merge(nef, s_m, left_index=True, right_index=True)

    s_m = s_m.rename(columns={'Polarity':'Positive', 'Polarity_x':'Neutral', 'Polarity_y':'Negative'})

    return s_m

def mood_verbosity():

    tdf['Verbosity'] = tdf['Tweets'].str.len()

    mood_v = pd.DataFrame({'Positive': tdf[tdf['Polarity'] > 0]['Verbosity'].mean(), 
                           'Neutral': tdf[tdf['Polarity'] == 0]['Verbosity'].mean(),
                           'Negative': tdf[tdf['Polarity'] < 0]['Verbosity'].mean()
                         }, index=[0])

    return mood_v

def popular_moods():

    pop_mood = pd.DataFrame({'Positive': len(tdf[tdf['Polarity'] > 0]),
                             'Neutral': len(tdf[tdf['Polarity'] == 0]),
                             'Negative': len(tdf[tdf['Polarity'] < 0])
                           }, index=[0])

    return pop_mood

def average_subjectivity():

    mood_subj = pd.DataFrame({'Positive': tdf[tdf['Polarity'] > 0]['Subjectivity'].mean(),
                           'Negative': tdf[tdf['Polarity'] < 0]['Subjectivity'].mean()
                         }, index=[0])

    mood_subj = mood_subj.abs()

    return mood_subj

tdf = get_df_from_db()
sent = sentiment_frames()
mdf = mood_verbosity()
pdf = popular_moods()
adf = average_subjectivity()
app.layout = html.Div(children=[
    html.Div(children=[
        html.H1(children='Climate Mood'),
        html.Div(children=['''
            Hourly Twitter sentiment analysis on climate change
        ''', 

            html.A(html.Button('Source Code!'),
                href='https://github.com/aakashhdesai/climate-mood',
            )
        ], className='description'),
    ], className='navigation'),

    html.Div(children=[

        html.Div(children=[
            html.H3(children=['Tweet Sentiment vs. Time (Per Min)']),
        ], className='senti-title'),
        dcc.Graph(
            id='sentiment-chart',
            figure={
                'data': [
                    go.Bar(
                        x=sent.index, y=sent['Negative'], name='Negative', marker={'color': '#bf5241'}
                    ),
                    go.Bar(
                        x=sent.index, y=sent['Neutral'], name='Neutral', marker={'color': '#417cbf'}
                    ),
                    go.Bar(
                        x=sent.index, y=sent['Positive'], name='Positive', marker={'color': '#5fbf41'}
                    ),
                ],
                'layout': go.Layout(
                    margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                    hovermode='closest',
                    barmode='stack',
                    legend=dict(
                        x=0,
                        y=1,
                        traceorder='normal',
                        font=dict(
                            size=12,
                            color='#000'
                        ),
                        bgcolor='#E2E2E2',
                        bordercolor='#FFFFFF',
                        borderwidth=2
                        )
                )
            }
        )
    ], className='sentiment'),

    html.Div(children=[
        html.Div(children=[
            html.H3(children=['''Mood Verbosity''']),
            html.Div(children=[
                dcc.Graph(
                    figure={
                        'data': [go.Bar(x=mdf.iloc[0],
                                        y=['Positive', 'Neutral', 'Negative'],
                                        orientation='h',
                                        text=mdf.iloc[0].round(2),
                                        textposition = 'auto',
                                        marker={'color': ['#5fbf41', '#417cbf', '#bf5241']})
                        ],
                        'layout': go.Layout(
                                  title='Avg Characters per Tweet',
                                  hovermode=False,
                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                  margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                                  width=300,
                                  height=300
                        )
                    })
            ], className='verbosity')
        ], className = 'analysis'),

        html.Div(children=[
            html.H3(children=['''Popular Moods''']),
            html.Div(children=[
                dcc.Graph(
                    figure={
                        'data': [go.Pie(values=pdf.iloc[0],
                                        labels=['Positive', 'Neutral', 'Negative'],
                                        text=pdf.iloc[0].round(2),
                                        marker={'colors': ['#5fbf41', '#417cbf', '#bf5241']})
                        ],
                        'layout': go.Layout(
                                  title='Sum of Tweets per Mood',
                                  hovermode=False,
                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                  margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                                  width=300,
                                  height=300,
                                  showlegend=False
                        )
                    })
            ], className='popular')
        ], className = 'analysis'),

        html.Div(children=[
            html.H3(children=['''Mood Rating''']),
            html.Div(children=[
                dcc.Graph(
                    figure={
                        'data': [go.Bar(y=adf.iloc[0],
                                        x=['Positive', 'Negative'],
                                        text=adf.iloc[0].round(2),
                                        textposition = 'auto',
                                        marker={'color': ['#49ad4b', '#ad4b49']})
                        ],
                        'layout': go.Layout(
                                  title='Avg Subjectivity per Mood',
                                  hovermode=False,
                                  xaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                  yaxis=dict(showgrid=False, zeroline=False, ticks='', showticklabels=False, fixedrange=True),
                                  margin={'l': 40, 'b': 40, 't': 40, 'r': 40},
                                  width=300,
                                  height=300
                        )
                    })
            ], className='mood-ratio')

        ], className = 'analysis')
    ], className='mood-stats')
])

if __name__ == '__main__':
    app.run_server(debug=True)
